gate-core/src/main/groovy/com/netflix/spinnaker/gate/security/RequestContext.java

```diff
@@ -16,61 +16,71 @@
 
 package com.netflix.spinnaker.gate.security;
 
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * No longer backed by a thread local, this is now a readonly proxy to {@link AuthenticatedRequest},
+ * which is the only source of truth for context values. This is to avoid the situation where
+ * RequestContext and {@link AuthenticatedRequest} provide an inconsistent view of the context or
+ * lose it across thread boundaries.
+ */
+@Slf4j
 public class RequestContext {
-  private static final ThreadLocal<RequestContext> threadLocal = new ThreadLocal<>();
+  private static final RequestContext placeholder = new RequestContext();
 
-  private String application;
-  private String authenticatedUser;
-  private String executionType;
-  private String executionId;
-  private String origin;
+  private RequestContext() {}
 
   public RequestContext(String origin, String authenticatedUser) {
-    this.origin = origin;
-    this.authenticatedUser = authenticatedUser != null ? authenticatedUser : "anonymous";
+    log.warn("call to deprecated RequestContext's constructor");
   }
 
+  @Deprecated
   public static void set(RequestContext requestContext) {
-    threadLocal.set(requestContext);
+    log.warn("call to deprecated method RequestContext::set");
   }
 
+  @Deprecated
   public static void setApplication(String application) {
-    threadLocal.get().application = application;
+    log.warn("call to deprecated method RequestContext::setApplication");
   }
 
+  @Deprecated
   public static void setExecutionType(String executionType) {
-    threadLocal.get().executionType = executionType;
+    log.warn("call to deprecated method RequestContext::setExecutionType");
   }
 
+  @Deprecated
   public static void setExecutionId(String executionId) {
-    threadLocal.get().executionId = executionId;
+    log.warn("call to deprecated method RequestContext::setExecutionId");
   }
 
-  public static RequestContext get() {
-    return threadLocal.get();
+  @Deprecated
+  public static void clear() {
+    log.warn("call to deprecated method RequestContext::clear");
   }
 
-  public static void clear() {
-    threadLocal.remove();
+  public static RequestContext get() {
+    return placeholder;
```

Comment:
is this still needed? if you remove won't that obviate the need to for this class all together (since the compiler (except groovy i guess) will find all the places it's used)?

Reply:
ah, just saw your other comment. makes sense with #3